### ISSUE TITLE
Fail when trying to use postgressql provider against aws rds instance in same definition

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-postgresql#2]
_This issue was originally opened by @chaliy as hashicorp/terraform#7442. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

I am trying to implement PostgreSQL database on AWS  RDS. I am crating PostgreSQL RDS Instance, and then database"

``` hcl
provider "aws" {}
resource "aws_db_instance" "main" {}

provider "postgresql" {}
resource "postgresql_database" "prototype" {}
```

Unfortunately this does not work. Provider postgresql for `plan` requires connection, and it is not yet here, because instance is not created.

```
postgresql_database.prototype: Error reading info about database: dial tcp: lookup ${aws_db_instance.address}: getaddrinfow: No such host is known.
```

or

```
cannot parse '' as int: strconv.ParseInt: parsing "": invalid syntax
```
### Expected Behavior

It was expected that somehow postgresql provider will figure out that instance is not here, and will not crash.
### Actual Behavior

Deployment crashes with error
### Steps to Reproduce

Please list the steps required to reproduce the issue, for example:
1. https://gist.github.com/chaliy/6af4c607afb04e356ca240f9df45d066
2. `terraform plan`

**Comment by sean-** [comment]
@chaliy , this won't exactly help the above issue, but I'd be curious as to your thoughts after reading the following:

https://github.com/terraform-providers/terraform-provider-postgresql/issues/5#issuecomment-314031307

The problem is that the underlying database connection pooler (`sql.DB`) doesn't allow for post-connection initialization hooks in the same way that `github.com/jackc/pgx` does.  #11 has been opened to track the migration to `jackc/pgx`.  Specifically, this nugget of functionality is required to solve this issue is:

https://github.com/jackc/pgx/blob/master/conn_pool.go#L12
https://godoc.org/github.com/jackc/pgx#ConnPoolConfig

I want to use `jackc/pgx` for a few reasons unrelated to your issue, but late-binding connections from the provider is one of them.  This will prevent `terraform plan` from attempting to establish new connections.

**Comment by tomelliff** [comment]
Is this still true? I thought that might be the issue when I started trying to have Terraform create an RDS instance and then add resources but it seems to plan and apply out okay from scratch for me.

I can provide an example of this working if it helps at all.

**Comment by sean-** [comment]
@tomelliff Is what still true?

**Comment by tomelliff** [comment]
That it fails to plan when using the postgres provider in the same Terraform config as the RDS instance and using the `aws_db_instance.resource_name.address` as the `host` for the `postgresql` provider details. Because it seems to work fine for me.

**Comment by sean-** [comment]
Oh, I think I understand what you're saying.  Namely that the PostgreSQL resource fails to plan because it has a dependency on the RDS instance being available, but if you attempt to mix and match the terraform to create the RDS instance with plan to create the schema, the provider fails to plan because it can't connect to the yet-to-be-created instance.

There are two possible outcomes here:

1) Create a provider-level dependency where the postgresql provider has a dependency on the rds provider (this doesn't exist yet).
2) Have the provider return an empty config suggesting everything needs to happen in the plan.

I don't like #2 because it could be a transient connectivity problem that would result in a plan being created that would attempt to recreate the database schema.  I think the current behavior, while not optimal, is correct for the time being.

The ideal solution would be for there to be a second-round of planning that would happen after the RDS instance is provisioned.  I'm not sure how to do that at plan time, however.  I need to think more about this, but that's my current thinking.

**Comment by tomelliff** [comment]
I was asking if the original poster still had issues because it works fine for me as is. Unfortunately the config they pasted is too minimal to reproduce it but as far as I can see this issue could be closed.

**Comment by sean-** [comment]
@tomelliff if you can post sample configs, that'd be great!

**Comment by gabro** [comment]
@tomelliff I'm facing the exact same issue as the OP, so if you could post your config it'd be great!

**Comment by tomelliff** [comment]
I was wrong apparently, plans before the RDS instance are created fail but you can apply out anyway and that manages the dependency order fine.

**Comment by gabro** [comment]
So you mean something like

```
terraform apply -target aws_rds_instance.my_instance
terraform apply
```

?

**Comment by tomelliff** [comment]
I've been considering requiring plans for every apply (and having CI pass the plan to the apply stage) but not overly sure if it's necessary because I'm not massively risk averse and have reasonable test environments before production. That said, if I did have this requirement then the inability to plan a database before applying it the first time would be a deal breaker. It's probably not an easy thing to fix though.

**Comment by gabro** [comment]
Ah, thanks for the pointers! I'll take a look

**Comment by bloo** [comment]
I'm seeing this during `plan` when running on Atlas (Terraform Enterprise), which runs on Ubuntu, but not locally! - on  my Darwin/OS X machine. Are there different versions of the provider plugin? I'll be watching this closely and will gladly provide more info if needed.

(fyi - version 0.11.2)

**Comment by sean-** [comment]
@bloo Can you paste your error message from your plan?

@tomelliff What do you have `"${var.database_port}"` set to?

**Comment by bloo** [comment]
@sean- I keep it as its default, but I just tried it set to `5432` and saw the same issue.

```
Error: Error refreshing state: 1 error(s) occurred:

* module.rds-pg.provider.postgresql: Error initializing PostgreSQL client: error detecting capabilities: error PostgreSQL version: dial tcp :5432: getsockopt: connection refused
```

**Comment by sean-** [comment]
@bloo in your run, can you set the logging level to `INFO` and re-run?  There should be a line that looks like:

```
[INFO] PostgreSQL DSN: `your DSN here`
```

that will help shed some light on this.

**Comment by bloo** [comment]
@sean- here's the output with `TF_LOG=INFO` set:

https://gist.github.com/bloo/638a769ac134bbff4d43974fadafbca5

**Comment by bloo** [comment]
@sean- any headway into this? Is there anything else I can provide?

**Comment by radeksimko** [comment]
Hi folks,
sorry this isn't working as expected.

As it was indirectly mentioned this is related to a known core issue where we perform the interpolation (of the DB hostname in this case) too early in the lifecycle.

See e.g. https://github.com/hashicorp/terraform/issues/12393 which was later consolidated to https://github.com/hashicorp/terraform/issues/4149

In regards to workarounds - some have been mentioned already:

### Separate out `aws` and `postgres` configs

Keep each in its own directory and `apply` each separately while leveraging the remote state to pass the data in between.

Example:

**`/aws/main.tf`**
```hcl
terraform {
  backend "s3" {
    bucket = "mybucket"
    key    = "path/to/my/key"
    region = "us-east-1"
  }
}

resource "aws_db_instance" "main" {
...
}

output "host" {
  value = "${aws_db_instance.main.address}"
}
output "port" {
  value = "${aws_db_instance.main.port}"
}
output "username" {
  value = "${aws_db_instance.main.username}"
}
output "password" {
  value = "${aws_db_instance.main.password}"
}
```

**`/postgres/main.tf`**
```hcl
data "terraform_remote_state" "aws" {
  backend = "s3"

  config {
    bucket = "mybucket"
    key    = "path/to/my/key"
    region = "us-east-1"
  }
}

provider "postgresql" {
  alias = "repro"
  host = "${data.terraform_remote_state.aws.host}"
  port = "${data.terraform_remote_state.aws.port}"
  username = "${data.terraform_remote_state.aws.username}"
  password = "${data.terraform_remote_state.aws.password}"
  sslmode = "require"
}

resource "postgresql_database" "repro1" {
  provider = "postgresql.repro"
  name = "repro1"
}

resource "postgresql_database" "repro2" {
  provider = "postgresql.repro"
  name = "repro2"
}
```

You can read more about remote state [here](https://www.terraform.io/docs/state/remote.html).

### Perform two-staged apply

```
terraform apply -target=aws_db_instance.main && terraform apply
```

The first `apply` will create the database instance, save all the details into state and the second apply will start at the point where the instance is already up with all necessary details (like hostname) stored in the state.

Read more about targeting [here](https://www.terraform.io/docs/commands/plan.html#resource-targeting).

-----

I'd recommend subscribing to the mentioned core issue if you want to hear about updates (i.e. any progress on resolving that issue).

**Comment by bloo** [comment]
Hi @radeksimko - we had hoped to not split our module up if the fix was going to happen soon(ish), but seeing as the the core issue is 2+ years old, do you recommend that we move forward with the split and consider this issue as the state of the world for now?

**Comment by radeksimko** [comment]
@bloo fix for the mentioned issue is not scheduled for the upcoming release (0.12) and I can't comment on the roadmap beyond that, so I'd recommend you do that.

Thanks for your understanding.

**Comment by bloo** [comment]
@radeksimko thanks for the transparency 👍

**Comment by tomelliff** [comment]
I noticed the other day that while I can't plan a config that has an RDS instance to be created plus any Postrgesql resources as others are trying I can apply it from scratch with the 0.1 provider version but this is broken in 0.1.1 (presumably due to the feature checking in https://github.com/terraform-providers/terraform-provider-postgresql/pull/12/files). Note that I'm also using Terraform 0.10.8 at the moment but am about to bump that to latest (0.11.4) and I'm unsure if the apply without a plan will work in 0.11 with the forced plan. I'm hoping `-auto-approve=true` should solve that

**Comment by mwz** [comment]
Hi, I'm experiencing the same issue, has there been any progress on this?

**Comment by scott2b** [comment]
Are these workarounds expected to work for an RDS instance inside a VPC? From what I can see, that is not the case. This problem persists for me despite having pre-created the instance and simply running a postgresql provider database and/or role create resource in an isolated terraform against the existing endpoint.

I have tried adding a NAT gateway, as well as adding a public subnet to the subnet group of the instance, but still no luck -- I get the same timeout issue, and am unsure of what else to try.

**Comment by tburow** [comment]
bump on this issue. still seeing this, im really surprised that this hasnt been resolved?

will this be resolved in the 12 release or?  ETA on the 12 release?

**Comment by cyrilgdn** [comment]
Hi,

As [very well explained](https://github.com/terraform-providers/terraform-provider-postgresql/issues/2#issuecomment-369341707) by @radeksimko, this is Terraform core issue. I recommend everyone concerned to read the aforementioned issue: https://github.com/hashicorp/terraform/issues/4149.

So I'm in favor to close this issue. Any opinion?

**Comment by cyrilgdn** [comment]
@tomelliff Thanks for the clarification. I just tested this case in 0.1.0 and I get your point. I'll try to check what can be done about that.

**Comment by milosgajdos83** [comment]
Silly question, but if the advice is to split your `aws` and `postgres` configs into two separate directories, how does one deal with multliple `init`s?

First, one needs to init the `aws` config (which generates local `.terraform` directory), then the `postgres` config requires to be intiialized too which if done from the "root" directory it will override the previously created `.terraform` directory with `postgres` config one.  Is there any terraform magic trick I'm missing?

**Comment by isen-ng** [comment]
My terraform scripts creates an RDS instance and attempts to use the postgres provider together in the same run. Fails on provider version 0.3.0, saying that it cannot connect to instance `:0` (because the RDS instance hasn't been created yet)

But I can do the same thing on provider version 0.1.0 smoothly
```
$ terraform plan
$ terraform apply
```

**Comment by so0k** [comment]
similar dependency issue with creating a schema and adding extensions into the schema, need to be split as well resulting in 3 folders:

- `aws/main.tf`
- `postgres/database.tf`
- `postgres/extensions.tf`

with a lot of microservices, this becomes unsustainable

**Comment by cyrilgdn** [comment]
@so0k Hmm this should not be the case for the db/schema/extension part. Only the `aws` part should be (currently) run in a separate step. Dependencies between Postgres resources should work.

If it's not the case could give an example with the associated error message please?

**Comment by outime** [comment]
I can't understand why the MySQL provider works fine with the same setup (I just tested it) but this provider can't, given that it's supposed to be a core Terraform issue. I'm not expecting a straight answer given that we're talking about completely different providers but I'm genuinely wondering.

**Comment by nunofernandes** [comment]
Also have the same issue after upgrading to version 1.0 of the module (and terraform 0.12).

**Comment by axmetishe** [comment]
The main problem here is that plugin tries to determine engine version on plugin configuration stage:

```go
// postgresql/provider.go:18
func Provider() terraform.ResourceProvider {

// postgresql/provider.go:111
		ConfigureFunc: providerConfigure,
	}
}
```

```go
// postgresql/provider.go:138
func providerConfigure(d *schema.ResourceData) (interface{}, error) {

// postgresql/provider.go:165
	client, err := config.NewClient(d.Get("database").(string))
```
```go
// postgresql/config.go:116
func (c *Config) NewClient(database string) (*Client, error) {

	if !found {
// postgresql/config.go:123
		db, err := sql.Open("postgres", dsn)
		if err != nil {
			return nil, errwrap.Wrapf("Error connecting to PostgreSQL server: {{err}}", err)
		}

// postgresql/config.go:134
		version, err := fingerprintCapabilities(db)
		if err != nil {
			db.Close()
			return nil, errwrap.Wrapf("error detecting capabilities: {{err}}", err)
		}
```
which a sql request to the database in the essence:
```go
// postgresql/config.go: 275
func fingerprintCapabilities(db *sql.DB) (*semver.Version, error) {
	var pgVersion string
	err := db.QueryRow(`SELECT VERSION()`).Scan(&pgVersion)
	if err != nil {
		return nil, errwrap.Wrapf("error PostgreSQL version: {{err}}", err)
	}
```

Engine version used for checking capabilities:
```go
var (
	dbRegistryLock sync.Mutex
	dbRegistry     map[string]dbRegistryEntry = make(map[string]dbRegistryEntry, 1)

	// Mapping of feature flags to versions
	featureSupported = map[featureName]semver.Range{
		// CREATE ROLE WITH
		featureCreateRoleWith: semver.MustParseRange(">=8.1.0"),

		// CREATE DATABASE has ALLOW_CONNECTIONS support
		featureDBAllowConnections: semver.MustParseRange(">=9.5.0"),

		// CREATE DATABASE has IS_TEMPLATE support
		featureDBIsTemplate: semver.MustParseRange(">=9.5.0"),

		// https://www.postgresql.org/docs/9.0/static/libpq-connect.html
		featureFallbackApplicationName: semver.MustParseRange(">=9.0.0"),

		// CREATE SCHEMA IF NOT EXISTS
		featureSchemaCreateIfNotExist: semver.MustParseRange(">=9.3.0"),

		// REASSIGN OWNED BY { old_role | CURRENT_USER
		featureReassignOwnedCurrentUser: semver.MustParseRange(">=9.5.0"),

		// row-level security
		featureRLS: semver.MustParseRange(">=9.5.0"),

		// CREATE ROLE has REPLICATION support.
		featureReplication: semver.MustParseRange(">=9.1.0"),

		// CREATE EXTENSION support.
		featureExtension: semver.MustParseRange(">=9.1.0"),

		// We do not support postgresql_grant and postgresql_default_privileges
		// for Postgresql < 9.
		featurePrivileges: semver.MustParseRange(">=9.0.0"),
	}
)
```

I have no idea how to handle this properly, because version checking is required for resource proper configuration, but at the same time we should not be pinned to instance existence on the 'plan' stage.

Currently I've worked this around using dirty(really dirty) hack - I've mocked instance version string and disable error checks on at providerConfigure method.
This allow me to make fake dependencies via local resources between modules and connect to RDS instance only after its creation.

Hope this helps to someone.

P.S. Sorry I'm a bit late, but Happy birthday, dear Bug=)

**Comment by nunofernandes** [comment]
I think the best way of handling this is to have a `depends_on = ["rds.instance"]` feature in the provider information. But for that, it would require that https://github.com/hashicorp/terraform/issues/2430 is implemented.

**Comment by axmetishe** [comment]
@nunofernandes I'm not sure that 2430 will resolve the issues because request to database generated at plan stage.
But may be switching module to handle compatibility property like compatibilityVersion does

**Comment by josegonzalez** [comment]
For the Postgres provider, could we possibly set the version in the provider to override the version lookup? I assume that would mean BadThingsForInvalidVersion™, but it would allow for folks who know the version ahead of time to do the right thing vs force partial applications.

**Comment by drpebcak** [comment]
This issue has been open since 2016 if you look back past the migration.. is this not considered a major issue with the provider? Not being able to create the db and then manage it is a huge problem (and by the way, the mysql provider does not have this issue).

**Comment by tburow** [comment]
I've ceased even trying to use this provider - in its current form its more or less useless.  I know thats probably not very constructive but we have defaulted back to sql scripts and null resources.....

**Comment by cyrilgdn** [comment]
@drpebcak I have an idea on how to fix that (or at least a patch I want to try).

I'll try to do it as soon as I can but I haven't so much time currently.

**Comment by jwilges** [comment]
@cyrilgdn is your fix still tied to the issue mentioned by @axmetishe (quoted below)?

> The main problem here is that plugin tries to determine engine version on plugin configuration stage

Is it feasible to just delay the version check until the `apply` stage? Or, are you proposing a different way to work through this issue?

Failing early during the `plan` stage seems like a "nice to have" _at most_ considering the responses here indicate users who already have an endpoint running will still hit the unavoidable early exit during `plan` unless they create another terraform repository just for the `postgresql` provider and apply it later.

**Comment by axmetishe** [comment]
@jwilges 
I know that there are should be a better options but for now I've ended up with dirty hack which allow me to provide all configuration within single module. Will leave it here:
```
diff --git a/postgresql/config.go b/postgresql/config.go
index d39fd2d..6e7a156 100644
--- a/postgresql/config.go
+++ b/postgresql/config.go
@@ -274,10 +274,11 @@ func (c *Client) DB() *sql.DB {
 // capabilities.  This is only run once per Client.
 func fingerprintCapabilities(db *sql.DB) (*semver.Version, error) {
        var pgVersion string
-       err := db.QueryRow(`SELECT VERSION()`).Scan(&pgVersion)
-       if err != nil {
-               return nil, errwrap.Wrapf("error PostgreSQL version: {{err}}", err)
-       }
+       pgVersion = "PostgreSQL 11.2 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.3 20140911 (Red Hat 4.8.3-9), 64-bit"
+       //err := db.QueryRow(`SELECT VERSION()`).Scan(&pgVersion)
+       //if err != nil {
+       //      return nil, errwrap.Wrapf("error PostgreSQL version: {{err}}", err)
+       //}
 
        // PostgreSQL 9.2.21 on x86_64-apple-darwin16.5.0, compiled by Apple LLVM version 8.1.0 (clang-802.0.42), 64-bit
        // PostgreSQL 9.6.7, compiled by Visual C++ build 1800, 64-bit
diff --git a/postgresql/provider.go b/postgresql/provider.go
index 9a0b1f1..f0f9883 100644
--- a/postgresql/provider.go
+++ b/postgresql/provider.go
@@ -4,7 +4,7 @@ import (
        "fmt"
 
        "github.com/blang/semver"
-       "github.com/hashicorp/errwrap"
+       //"github.com/hashicorp/errwrap"
        "github.com/hashicorp/terraform/helper/schema"
        "github.com/hashicorp/terraform/terraform"
 )
@@ -163,11 +163,11 @@ func providerConfigure(d *schema.ResourceData) (interface{}, error) {
        }
 
        client, err := config.NewClient(d.Get("database").(string))
-       if err != nil {
-               return nil, errwrap.Wrapf("Error initializing PostgreSQL client: {{err}}", err)
-       }
+       //if err != nil {
+       //      return nil, errwrap.Wrapf("Error initializing PostgreSQL client: {{err}}", err)
+       //}
 
-       return client, nil
+       return client, err
 }
 
 func tfAppName() string {
```

**Comment by cyrilgdn** [comment]
Hi all,

Thanks to @vboginskey patch in https://github.com/terraform-providers/terraform-provider-postgresql/pull/103, there's now a simple workaround for this problem. 

You can specify the `expected_version` attribute in your provider configuration with the engine version.

i.e.:
```
resource "aws_db_instance" "test_db" {
    [...]
}

provider "postgresql" {
  host             = aws_db_instance.test_db.address
  port             = aws_db_instance.test_db.port
  username         = aws_db_instance.test_db.username
  password         = var.postgres_password
  expected_version = aws_db_instance.test_db.engine_version
  sslmode          = "require"
}
```

The provide will not try to detect the version and you'll be able to plan/apply both resources at the same time.

**Comment by cyrilgdn** [comment]
FYI, @vboginskey 's patch has been released  in [v1.4.0](https://github.com/terraform-providers/terraform-provider-postgresql/releases/tag/v1.4.0) 2 days ago !

**Comment by dishbreak** [comment]
Quick addition to future searchers. If you are setting an alias on your provider and forget to put the provider on a resource, you'll run into this problem. I fixed it by making sure that all resources had the `provider` named appropriately.

Why is this a problem? I took a look at debug logs, and I saw this behavior:

```
2020/08/06 12:06:15 [DEBUG] adding implicit provider configuration provider.postgresql, implied first by ...
```

When I looked at the resource in question, I realized the `provider` block was missing and a new one was being generated that didn't use my config!

**Comment by jakubigla** [comment]
This still doesn't work for me (I'm on v1.7.1)
While the first plan and apply works, the subsequent one fails...

**Comment by qlikcoe** [comment]
We're experiencing the same issue when trying to create a PostgreSQL RDS using `terraform-aws-modules/rds/aws` module and then provisioning stuff with `PostgreSQL` provider. The first run is fine, the subsequent runs fail with `Error: dial tcp :5432: connect: connection refused`. I'm assuming the hostname is empty because of some racing conditions and reasons mentioned above in this thread. 

The workaround I've come out with is to use an intermediate SSM secret for the hostname. So During the first run the hostname is written as a Parameter Store secret. For the subsequent `terraform apply` the postgresql provider is using that hostname secret and it doesn't fail anymore. 

This is the new resource I've added
```
resource "aws_ssm_parameter" "cortex-db-hostname" {
  name        = "/cortex/db_hostname"
  type        = "SecureString"
  description = "Cortex DB Hostname"
  value       = module.db.this_db_instance_address
  overwrite   = true
}
```

and then the provider config looks like this: 

```
provider "postgresql" {
  host            = aws_ssm_parameter.cortex-db-hostname.value
  port            = var.db_port
  username        = var.db_root_user
  password        = var.db_root_password
  superuser       = false
  sslmode         = "require"
  connect_timeout = 15
  expected_version = "12.4"
}
```

**Comment by jakubigla** [comment]
@qlikcoe I think this issue is caused by some "module" internals. So the workaround I applied is below:
`//workaround for postgres provider
resource "null_resource" "test" {
  triggers = {
    host     = module.geodb_rds.this_db_instance_address
    port     = module.geodb_rds.this_db_instance_port
    username = module.geodb_rds.this_db_instance_username
    dbname   = module.geodb_rds.this_db_instance_name
  }
}`

And then in the postgres provider configuration I do this:
`provider "postgresql" {
  host     = null_resource.test.triggers.host
  port     = null_resource.test.triggers.port
  database = null_resource.test.triggers.dbname
  username = null_resource.test.triggers.username
  password = data.aws_secretsmanager_secret_version.geodb_rds_password.secret_string

  expected_version = "12.3"

  sslmode   = "require"
}`

Hope it will get you going while I know it's not ideal. ¯\_(ツ)_/¯
